packages/opencensus-web-instrumentation-zone/src/perf-resource-timing-selector.ts

```diff
@@ -0,0 +1,169 @@
+ * overlap timings with other resource timing entry. Also, a possible entry
+ * without cors resource timing is when that resource timing entry is not
+ * 'paired' with any other entry.
+ * Thus, for this step traverse the array of resource entries and for every
+ * entry check if it is a possible performance resource entry.
+ * @param perfEntries
+ */
+export function getPossiblePerfResourceEntries(
+  perfEntries: PerformanceResourceTiming[]
+): XhrPerformanceResourceTiming[] {
+  const possiblePerfEntries = new Array<XhrPerformanceResourceTiming>();
+  const pairedEntries = new Set<PerformanceResourceTiming>();
+  let entryI: PerformanceResourceTiming;
+  let entryJ: PerformanceResourceTiming;
+  // As this part of the algorithm traverses the array twice, although,
+  // this array is not large as the performance resource entries buffer is
+  // cleared when there are no more running XHRs.
+  for (let i = 0; i < perfEntries.length; i++) {
+    entryI = perfEntries[i];
+    // Compare every performance entry with its consecutive perfomance entries.
+    // That way to avoid comparing twice the entries.
+    for (let j = i + 1; j < perfEntries.length; j++) {
+      entryJ = perfEntries[j];
+      if (!overlappingPerfResourceTimings(entryI, entryJ)) {
+        // As the entries are not overlapping, that means those timings
+        // are possible perfomance timings related to the XHR.
+        possiblePerfEntries.push({
+          corsPreFlightRequest: entryI,
+          mainRequest: entryJ,
+        });
+        pairedEntries.add(entryI);
+        pairedEntries.add(entryJ);
+      }
+    }
+    // If the entry couldn't be paired with any other resource timing,
+    // add it as a possible resource timing without cors preflight data.
+    // This is possible because this entry might be better than the other
+    // possible entries.
+    if (!pairedEntries.has(entryI)) {
+      possiblePerfEntries.push({ mainRequest: entryI });
+    }
+  }
+  return possiblePerfEntries;
+}
+
+// Pick the best performance resource timing for the XHR: Using the possible
+// performance resource timing entries from previous step, the best entry will
+// be the one with the minimum gap to the span start/end timings.
+// The performance resource timing entry with the minimum gap to the span
+// start/end timings points out that entry is the best fit for the span.
+function getBestPerfResourceTiming(
+  perfEntries: XhrPerformanceResourceTiming[],
+  span: Span
+): XhrPerformanceResourceTiming | undefined {
+  let minimumGapToSpan = Number.MAX_VALUE;
+  let bestPerfEntry: XhrPerformanceResourceTiming | undefined = undefined;
+  for (const perfEntry of perfEntries) {
+    let gapToSpan = Math.abs(
+      perfEntry.mainRequest.responseEnd - span.endPerfTime
+    );
+    // If the current entry has cors preflight data use its `startTime` to
+    // calculate the gap to the span.
+    if (perfEntry.corsPreFlightRequest) {
```

Comment:
This is ok as a conditional, but could be a bit shorter with something like:

const perfStart = perfEntry.corsPreFlightRequest ? perfEntry.corsPreFlightRequest.startTime : perfEntry.mainRequest.startTime;
gapToSpan += perfStart - span.startPerfTime